### PR TITLE
[angular] Add missing definitions for $q.race

### PR DESCRIPTION
**Pull request by thewarpaint**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://docs.angularjs.org/api/ng/service/$q
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by calebstdenis** [comment]
@thewarpaint Seems to be an issue with the build, not your changes. Sit tight for now.

File: types/angular/index.d.ts [hunk 1]
```diff
@@ -1,6 +1,9 @@
 // Type definitions for Angular JS 1.6
 // Project: http://angularjs.org
-// Definitions by: Diego Vilar <http://github.com/diegovilar>, Georgii Dolzhykov <http://github.com/thorn0>, Caleb St-Denis <https://github.com/calebstdenis>
+// Definitions by: Diego Vilar <http://github.com/diegovilar>
+//                 Georgii Dolzhykov <http://github.com/thorn0>
+//                 Caleb St-Denis <https://github.com/calebstdenis>
+//                 Eduardo Garcia <https://github.com/thewarpaint>
```
Comment (thorn0) [inline types/angular/index.d.ts]: These things are parsed by the bot to notify the reviewers, so this is essentially a request to become a reviewer. 1) Did you mean it? 2) You changed the format of the list. Make sure the bot understands it. I have doubts about it.

File: types/angular/index.d.ts [hunk 2]
```diff
@@ -1034,6 +1037,31 @@ declare namespace angular {
...
+        race<T1, T2>(values: [T1 | IPromise<T1>, T2 | IPromise<T2>]): IPromise<[T1, T2]>;
+        race<TAll>(promises: Array<IPromise<TAll>>): IPromise<TAll[]>;
+
+        /**
+         * Returns a promise that resolves or rejects as soon as one of those promises resolves or rejects, with the value or reason from that promise.
+         *
+         * @param promises A hash of promises.
+         */
+        race(promises: { [id: string]: IPromise<any>; }): IPromise<{ [id: string]: any; }>;
+        race<T extends {}>(promises: { [id: string]: IPromise<any>; }): IPromise<T>;
```
Comment (thorn0) [inline types/angular/index.d.ts]: `extends {}` looks weird. `{}` is an empty structure, so any type extends it by default. It should be possible to use mapped types for this definition, but this requires TypeScript 2.1. I'm not sure if we want to stay compatible with the older versions.

Comment (andy-ms) [inline types/angular/index.d.ts]: Shouldn't the argument type by `{ [id: string]: IPromise<T>; }`?

File: types/angular/index.d.ts [hunk 3]
```diff
@@ -1034,6 +1037,31 @@ declare namespace angular {
          * Creates a Deferred object which represents a task which will finish in the future.
          */
         defer<T>(): IDeferred<T>;
+
+        /**
+         * Returns a promise that resolves or rejects as soon as one of those promises resolves or rejects, with the value or reason from that promise.
+         *
+         * @param promises An array of promises.
+         */
+        race<T1, T2, T3, T4, T5, T6, T7, T8, T9, T10>(values: [T1 | IPromise<T1>, T2 | IPromise<T2>, T3 | IPromise<T3>, T4 | IPromise <T4>, T5 | IPromise<T5>, T6 | IPromise<T6>, T7 | IPromise<T7>, T8 | IPromise<T8>, T9 | IPromise<T9>, T10 | IPromise<T10>]): IPromise<[T1, T2, T3, T4, T5, T6, T7, T8, T9, T10]>;
```
Comment (andy-ms) [inline types/angular/index.d.ts]: Does it really make sense to use `T | Promise<T>` here? Would `race` just immediately return that value then?

File: types/angular/index.d.ts (hunk 1 above)
Comment (andy-ms) [inline types/angular/index.d.ts]: `dt-bot`'s been having trouble lately but we do intend to support this style.

File: types/angular/index.d.ts (hunk 2 above)
Comment (thorn0) [inline types/angular/index.d.ts]: As for mapped type, this works as expected in the [Playground](http://www.typescriptlang.org/play/#src=%0D%0Adeclare%20function%20race%3CT%3E%28promises%3A%20%7B%0D%0A%20%20%20%20%5BP%20in%20keyof%20T%5D%3A%20T%5BP%5D%20%7C%20Promise%3CT%5BP%5D%3E%0D%0A%7D%29%3A%20Promise%3CT%5Bkeyof%20T%5D%3E%3B%0D%0A%0D%0Aconst%20a%20%3D%20race%28%7B%0D%0A%20%20%20%20foo%3A%20Promise.resolve%281%29%2C%0D%0A%20%20%20%20bar%3A%20Promise.resolve%28'2'%29%2C%0D%0A%20%20%20%20qux%3A%20true%0D%0A%7D%29%3B%0D%0A)
```ts
race<T>(promises: {
    [P in keyof T]: T[P] | Promise<T[P]>
}): Promise<T[keyof T]>;
```
But [when I tried it a weak ago](https://gitter.im/Microsoft/TypeScript?at=5928b697cb83ba6a41021e4b), it didn't work. The current TS version there is 2.3.1-insiders.20170416. Would be good to know what it was back then.

Comment (andy-ms) [inline types/angular/index.d.ts]: There's really no need for a mapped type here though -- the keys don't matter, just that they all share a value type in common.

Comment (thorn0) [inline types/angular/index.d.ts]: They don't have to share the type. If we use the mapped type as I wrote above, the type of the return value is a union. Which is nice, but seems to not work with some not so old TS versions. I wish there was a preprocessor to support different typing for different TS versions.

**Comment by yuit** [comment]
Thanks for your contribution. This PR has unaddressed comment and cannot be merged in at the time being. For housekeeping purposes we are closing stale PRs. If you'd still like to merge this code in, please open a new PR that has been merged and rebased with the `master` branch.

**Pull request closed without merging**